### PR TITLE
Fix testlab timeout limit exceeded

### PR DESCRIPTION
**Pull request by lszucs**
- modified description to reflect TestLab max timeout changes
- introduced validation for `test_timeout`

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -26,6 +27,11 @@ import (
 	"github.com/bitrise-tools/go-steputils/tools"
 )
 
+const (
+	maxTimeout = 1800
+	timeoutPattern = `^\d+$|^\d+\.\d{1,9}$` // see: https://firebase.google.com/docs/test-lab/reference/testing/rest/v1/projects.testMatrices#testspecification
```
Comment (richard-bogdan-bitrise) [inline main.go]: As I read the specs, I found this:
A duration in seconds with up to nine fractional digits, terminated by 's'. Example: "3.5s".
I am missing the s from the end of the regex.

File: main.go [hunk 2]
```diff
@@ -26,6 +27,11 @@ import (
 	"github.com/bitrise-tools/go-steputils/tools"
 )
 
+const (
+	maxTimeout = 1800
```
Comment (richard-bogdan-bitrise) [inline main.go]: Please add the unit to the name as it is more readable.

File: main.go [hunk 3]
```diff
@@ -119,6 +125,19 @@ func (configs ConfigsModel) validate() error {
 	if err := input.ValidateIfPathExists(configs.ZipPath); err != nil {
 		return fmt.Errorf("Issue with ZipPath: %s", err)
```
Comment (richard-bogdan-bitrise) [inline main.go]: Error strings should not start with capital letters. Please apply this to every occurrence in the file.

File: main.go [hunk 4]
```diff
@@ -10,6 +10,7 @@ import (
 	"net/http"
 	"os"
 	"path/filepath"
+	"regexp"
 	"strconv"
 	"strings"
 	"text/tabwriter"
```
Comment (richard-bogdan-bitrise) [inline main.go]: I know it's not your code modification, but could you please remove the alias testing from the imports, as it is unneeded.

File: main.go [hunk 5]
```diff
@@ -215,8 +234,16 @@ func main() {
 			testModel.EnvironmentMatrix.IosDeviceList.IosDevices = append(testModel.EnvironmentMatrix.IosDeviceList.IosDevices, &newDevice)
 		}
 
+		timeout := configs.TestTimeout
```
Comment (richard-bogdan-bitrise) [inline main.go]: Please move the validation to a new method, to increase readability and testability.

File: main.go [hunk 6]
```diff
@@ -119,6 +125,19 @@ func (configs ConfigsModel) validate() error {
...
 		return fmt.Errorf("Issue with ZipPath: %s", err)
 	}
+	if err := validateTimeout(configs.TestTimeout); err != nil {
+		return fmt.Errorf("Issue with TestTimeout: %s", err)
+	}
+
+	return nil
+}
+
+func validateTimeout(duration string) error {
```
Comment (richard-bogdan-bitrise) [inline main.go]: I think naming is confusing, I would expect a method that checks if the given timeout is valid or not, but this only check if the given input has the correct format, so I recommend to rename this to something more specific, like validateTimeoutFormat

File: step.yml [hunk 7]
```diff
@@ -105,9 +105,9 @@ inputs:
       category: "Debug"
       title: "Test timeout"
       summary: |
-        The max time this test execution can run before it is cancelled. It does not include any time necessary to prepare and clean up the target device. The maximum possible testing time is 3600 seconds.
+        The max time this test execution can run before it is cancelled. It does not include any time necessary to prepare and clean up the target device. The maximum possible testing time is 1800 seconds.
```
Comment (richard-bogdan-bitrise) [inline step.yml]: Summary and Description are the same, maybe we should use the first sentence as the summary and the following as the description.

File: main.go (hunk 1 above)
Comment (lszucs) [reply main.go]: Very good catch! However, the `s` is, in fact, appended to the value at a later point.

https://github.com/bitrise-steplib/steps-virtual-device-testing-for-ios/blob/master/main.go#L219

File: main.go (hunk 3 above)
Comment (lszucs) [reply main.go]: You're right, in general, but in this case the error message is the 'top-level' error and is used by the `failf` function in main. Our logging convention is to start with a capital letter in these cases and use all lowercase in the 'embedded' error messages.

File: main.go (hunk 5 above)
Comment (trapacska) [inline main.go]: please move this validation part to the input validations, as if this input contains something wrong then we should fail ASAP, and also it makes more sense there maybe

File: main.go [hunk 8]
```diff
@@ -215,8 +234,16 @@ func main() {
...
+		if val, err := strconv.ParseFloat(timeout, 64); err != nil {
+			failf("could not parse float from timeout value (%s): %s", timeout, err)
+		} else if val > float64(maxTimeout) {
+			log.Warnf("timeout value (%f) is greater than available maximum (%f). Maximum will be used instead.", val, maxTimeout)
+			timeout = strconv.Itoa(maxTimeout)
+		}
+
 		testModel.TestSpecification = &testing.TestSpecification{
-			TestTimeout: fmt.Sprintf("%ss", configs.TestTimeout),
+			TestTimeout: fmt.Sprintf("%ss", timeout),
```
Comment (trapacska) [inline main.go]: what about `timeout+"s"`?

**Comment by trapacska** [comment]
🍌

**Pull request merged**